composer/callbacks/generate.py

```diff
@@ -0,0 +1,116 @@
+# Copyright 2022 MosaicML Composer authors
+# SPDX-License-Identifier: Apache-2.0
+
+# Copyright 2022 MosaicML LLM Foundry authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Periodically log generations to wandb from a set of prompts."""
+from typing import Any, List, Union, cast
+
+from torch.nn.parallel import DistributedDataParallel
+from torch.utils.data import DataLoader, TensorDataset
+
+from composer.callbacks.utils import create_interval_scheduler
+from composer.core import Callback, Event, State, Time, get_precision_context
+from composer.loggers import Logger
+from composer.models import HuggingFaceModel
+from composer.utils import dist
+from composer.utils.import_helpers import MissingConditionalImportError
+
+
+class Generate(Callback):
+    """Periodically log generations from a set of prompts.
+
+    Args:
+        prompts (List[str]): The list of prompts you would like to produce generations for
+        interval (Union[str, int, :class:`.Time`]): The interval describing how often checkpoints should be
+            saved. If an integer, it will be assumed to be in :attr:`.TimeUnit.EPOCH`.
+            Otherwise, the unit must be either :attr:`.TimeUnit.EPOCH`, :attr:`.TimeUnit.BATCH`,
+            :attr:`.TimeUnit.TOKEN`, or :attr:`.TimeUnit.SAMPLE`.
+        batch_size (int): Size of a prompt batch for generation.
+        kwargs: All kwargs well be passed along to the call to generate. This is for things like `do_sample`, `top_p`, etc
+    """
+
+    def __init__(self, prompts: List[str], interval: Union[str, int, Time], batch_size: int, **kwargs: Any):
+        self.prompts = prompts
+        self.generate_kwargs = kwargs
+        self.batch_size = batch_size
+        self.check_interval = create_interval_scheduler(interval, include_end_of_training=False)
+
+    def run_event(self, event: Event, state: State, logger: Logger) -> None:
+        if state.get_elapsed_duration() is not None and self.check_interval(state, event):
+            self.generate(state, logger)
+        else:
+            super().run_event(event, state, logger)
```

Comment:
Hm probably not for our particular case, but I figured it would be good to default to the original behavior, such that the other Callback behaviors work. For example, if someone wanted to extend Generate by overriding `after_load`, it would work as expected. wdyt?

Reply:
to really preserve behavior, i think `super().run_event(...)` should actually run always, not just in an `else` block

Reply:
I think the idea with callbacks is that you _either_ override `run_event` to do whatever you want, _or_ define the functions (e.g. `after_load`) and let `run_event` handle dispatching to them. So imo better to not mix the two methods and just override `run_event` here _or_ define the `batch_checkpoint` and `epoch_checkpoint` function calls like the checkpoint saver does. I don't have a strong opinion one way or the other, but I do think it is better not to mix the two methods.

Reply:
gotcha, sgtm!